### PR TITLE
RUMM-155 API Surface plugin

### PR DESCRIPTION
**Pull request by xgouchet**
### What does this PR do?

- [X] Create a plugin to generate the apiSurface definition
- [X] Add a task to detect whether apiSurface was updated in the PR
- [X] Add a label automatically to notify that the api surface changed

### Motivation

Have a way to detect early breaking changes for users

### Review checklist (to be filled by reviewers)

- [X] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [X] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [X] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

### CODE REVIEW

File: buildSrc/src/main/kotlin/com/datadog/gradle/plugin/apisurface/KotlinFileVisitor.kt [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+        description.append("\n")
+    }
+
+    private fun visitConstructor(constructor: Node.Decl.Constructor) {
+        description.append("constructor(")
+        visitParameters(constructor.params)
+        description.append(")\n")
+    }
+
+    private fun visitStructured(
```
Comment (mariusc83) [inline buildSrc/src/main/kotlin/com/datadog/gradle/plugin/apisurface/KotlinFileVisitor.kt]: Can you extract some functions and make this function a bit more readable please. It is very hard to follow now given its size.

File: buildSrc/src/test/kotlin/com/datadog/gradle/plugin/apisurface/KotlinFileVisitorTest.kt [hunk 2]
```diff
@@ -0,0 +1,631 @@
...
+
+import java.io.File
+import java.lang.IllegalStateException
+import junit.framework.Assert.assertEquals
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.rules.TemporaryFolder
+
+internal class KotlinFileVisitorTest {
```
Comment (mariusc83) [inline buildSrc/src/test/kotlin/com/datadog/gradle/plugin/apisurface/KotlinFileVisitorTest.kt]: Cool ;)

File: .github/labeler.yml [hunk 3]
```diff
@@ -0,0 +1,3 @@
+# Add 'api-surface-change' to any changes on the 'apiSurface' file in one of the library modules
+api-surface-change:
+  - dd-sdk-android/apiSurface
```
Comment (mariusc83) [inline .github/labeler.yml]: Can you be more specific here in what the _labeler.yml_ is doing ?  Meaning what it is going to execute and how it is linked with the _labels.yml_. Maybe a link to a Git docs will be better.

File: .github/workflows/labels.yml [hunk 4]
```diff
@@ -0,0 +1,12 @@
+name: Labeler
+on: [pull_request]
+
+jobs:
+  label:
+
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/labeler@v2
```
Comment (ncreated) [inline .github/workflows/labels.yml]: Where are those labels configured 🤔?

Screenshot 2020-01-15 at 10 12 04

Comment (ncreated) [reply .github/workflows/labels.yml]: I'd also say that `api-non-breaking-change` is too much. As most of PRs will be labelled with this one it's introducing some noise. Maybe we can add only `api-breaking-change` when needed?

File: .gitlab-ci.yml [hunk 5]
```diff
@@ -44,7 +44,15 @@ analysis:licenses:
   stage: analysis
   script:
     - git fetch --depth=1 origin master
-    - ./gradlew :dd-sdk-android:checkThirdPartyLicences
+    - ./gradlew :dd-sdk-android:checkThirdPartyLicences :dd-sdk-android-timber:checkThirdPartyLicences
+
+analysis:api-surface:
+  tags: [ "runner:docker", "size:large" ]
```
Comment (ncreated) [inline .gitlab-ci.yml]: What does `size:large` do in terms of picking the right runner?

File: .gitlab-ci.yml [hunk 6]
```diff
@@ -44,7 +44,15 @@ analysis:licenses:
...
+  image: $CI_IMAGE_DOCKER
+  stage: analysis
+  script:
+    - git fetch --depth=1 origin master
```
Comment (ncreated) [inline .gitlab-ci.yml]: This might keep giving false-positives on long-living feature stabilisation branches (like tracing). How about using `CI_MERGE_REQUEST_TARGET_BRANCH_NAME` GitLab variable?

File: buildSrc/src/main/kotlin/com/datadog/gradle/plugin/apisurface/GenerateApiSurfaceTask.kt [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+package com.datadog.gradle.plugin.apisurface
+
+import java.io.File
+import org.gradle.api.DefaultTask
+import org.gradle.api.tasks.InputDirectory
+import org.gradle.api.tasks.OutputFile
+import org.gradle.api.tasks.TaskAction
+
+open class GenerateApiSurfaceTask : DefaultTask() {
```
Comment (ncreated) [inline buildSrc/src/main/kotlin/com/datadog/gradle/plugin/apisurface/GenerateApiSurfaceTask.kt]: Is there a way to add unit tests for Gradle plugins' code? If yes, maybe it's worth adding at least integration test for this.

File: buildSrc/src/main/kotlin/com/datadog/gradle/plugin/apisurface/CheckApiSurfaceChangesTask.kt [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+        if (additions > 0 || removals > 0) {
+            throw IllegalStateException(
+                "Make sure you run the ${ApiSurfacePlugin.TASK_GEN_API_SURFACE} task before you push your PR."
+            )
+        }
+    }
+
+    @InputFile
+    fun getInputFile() = surfaceFile
```
[the remark this answers is not in the document]
Comment (xgouchet) [reply buildSrc/src/main/kotlin/com/datadog/gradle/plugin/apisurface/CheckApiSurfaceChangesTask.kt]: Yes, the `@InputFile` annotation lets gradle know if the task needs to be recomputed again. That way if the input didn't change, the task wont be ran again.

File: .github/workflows/labels.yml (hunk 4 above)
Comment (xgouchet) [reply .github/workflows/labels.yml]: The labels can be configured manually in the Github project directly.

File: .gitlab-ci.yml (hunk 5 above)
Comment (xgouchet) [reply .gitlab-ci.yml]: Not sure on the specifics, but the idea is that it will spawn the container in a runner that is made for large tasks.

File: .gitlab-ci.yml (hunk 6 above)
Comment (xgouchet) [reply .gitlab-ci.yml]: Because we don't have this value available. Git lab runs on commits and tags only, it's never triggered on pull requests.

File: buildSrc/src/main/kotlin/com/datadog/gradle/plugin/apisurface/GenerateApiSurfaceTask.kt (hunk 7 above)
Comment (xgouchet) [reply buildSrc/src/main/kotlin/com/datadog/gradle/plugin/apisurface/GenerateApiSurfaceTask.kt]: Testing task is not trivial, and most of the work is delegated to the KotlinFileVisitor anyway.

File: .github/labeler.yml (hunk 3 above)
Comment (xgouchet) [reply .github/labeler.yml]: done

File: buildSrc/src/main/kotlin/com/datadog/gradle/plugin/apisurface/CheckApiSurfaceChangesTask.kt [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    @TaskAction
+    fun applyTask() {
+        val lines = project.execShell(
+            "git", "diff", "--color=never", "HEAD", "--", surfaceFile.absolutePath
+        )
+
+        val additions = lines.count { it.matches(Regex("^\\+[^+].*$")) }
+        val removals = lines.count { it.matches(Regex("^-[^-].*$")) }
+
+        if (additions > 0 || removals > 0) {
```
Comment (mariusc83) [inline buildSrc/src/main/kotlin/com/datadog/gradle/plugin/apisurface/CheckApiSurfaceChangesTask.kt]: This task already depends on the TASK_GEN_API_SURFACE meaning that if there's any change in the public api the _surfaceFile_ will change meaning that this condition:
`if(additions > 0 || removals > 0) ....` will be true and throw ISE. Is that what you want, for me it's a bit weird.

File: buildSrc/src/main/kotlin/com/datadog/gradle/plugin/apisurface/KotlinFileVisitor.kt (hunk 1 above)
Comment (xgouchet) [reply buildSrc/src/main/kotlin/com/datadog/gradle/plugin/apisurface/KotlinFileVisitor.kt]: Done

File: buildSrc/src/main/kotlin/com/datadog/gradle/plugin/apisurface/CheckApiSurfaceChangesTask.kt (hunk 9 above)
Comment (xgouchet) [reply buildSrc/src/main/kotlin/com/datadog/gradle/plugin/apisurface/CheckApiSurfaceChangesTask.kt]: No, the idea is to make sure that the apiSurface file was properly generated with the commit(s) in the branch.
So if the apiSurface is in sync with what's in the source code in the same commit, the newly generated file will be identical so there'll be no addition nor removals
